### PR TITLE
[publictransportswitzerland] Public Transport Switzerland Initial contribution

### PR DESCRIPTION
**Pull request by jeremystucki**
Continuation of #7524

# Public Transport Switzerland Binding

Connects to the "Swiss public transport API" to provide real-time public transport information. [Link to the API](https://transport.opendata.ch/)

For example, here is a station board in HABPanel. (Download [here](https://github.com/StefanieJaeger/HABPanel-departure-board))

### Stationboard

Upcoming departures for a single station.

## Discovery

This binding does not support auto-discovery.

### Stationboard

`Station` is the station name for which to display departures.

### Stationboard

| channel | type   | description |
|---------|--------|-------------|
| departures#n   | String | A dynamic channel for each upcoming departure |
| tsv (advanced) | String | A tsv which contains the fields:<br />`identifier, departureTime, destination, track, delay` |

### CODE REVIEW

**Comment by fwolter** [comment]
Is this a draft by intention? If you like to get this reviewed, you should mark it as open.

**Comment by jeremystucki** [comment]
@fwolter It's a draft on purpose. I still have to address a review from the old PR. I just didn't have the time

**Comment by jeremystucki** [comment]
@fwolter Is it still possible to create dynamic channels from code? I don't see them appearing in my OH3 instance. I use `updateThing(editThing().withChannel(channel).build())`

**Comment by fwolter** [comment]
It should, but I never tried it with OH3.

**Comment by jeremystucki** [comment]
I can't seem to figure out why the dynamic channels don't show up in the UI.
If I look at `getThing().getChannelsOfGroup` they exist.

The static channel works as expected.

2021-02-14-154803_784x409_scrot

**Comment by fwolter** [comment]
I have no idea either. You could ask in the developer section in the forum.

**Comment by Skinah** [comment]
I have troubles with dynamic channels on the opensprinkler binding. This may be a bug in the core? as you can define them in the XML and remove them dynamically, but I could not add them.. I looked at other bindings for examples are could not see a reason why it should not work. Since it was only 1 binding that I did not write, I did not report it as a bug, but if your also having issues then it would pay to bring this up.

**Comment by jeremystucki** [comment]
I had another go at this and figured out that I had to define a channel type for the dynamic channels to show up in OH3.



Feel free to try out this plugin out or to review my code. I consider it to be fully migrated now.

File: bundles/org.openhab.binding.publictransportswitzerland/.classpath [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fwolter) [inline bundles/org.openhab.binding.publictransportswitzerland/.classpath:1]: This file is generated automatically and should not be part of your PR.

File: bundles/org.openhab.binding.publictransportswitzerland/.project [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fwolter) [inline bundles/org.openhab.binding.publictransportswitzerland/.project:1]: See above. Generated. Please remove.

File: bundles/org.openhab.binding.publictransportswitzerland/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+For example, here is a station board in HABPanel. (Download [here](https://github.com/StefanieJaeger/HABPanel-departure-board))
+
+![Departure board in HABPanel](doc/departure_board_habpanel.png)
+
+## Supported Things
+
+### Stationboard
+
+Upcoming departures for a single station. This is what you would usually see displayed at the train station.
```
Comment (fwolter) [inline bundles/org.openhab.binding.publictransportswitzerland/README.md:13]: You could mention the ThingTypeUID. You'll need it when using textual configuration. There are many users who use textual configuration.

File: bundles/org.openhab.binding.publictransportswitzerland/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+#### Configuration 
+
+`Station` is the station name for which to display departures.
```
Comment (fwolter) [inline bundles/org.openhab.binding.publictransportswitzerland/README.md:17]: The config parameter names are case sensitive.
```suggestion
`station` is the station name for which to display departures.  
```

File: bundles/org.openhab.binding.publictransportswitzerland/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+#### Channels 
+
+| channel | type   | description |
+|---------|--------|-------------|
+| departures#n   | String | A dynamic channel for each upcoming departure |
+| stationboard#tsv (advanced) | String | A tsv which contains the fields:<br />`identifier, departureTime, destination, track, delay` |
+
+## Discovery
+
+This binding does not support auto-discovery.
```
Comment (fwolter) [inline bundles/org.openhab.binding.publictransportswitzerland/README.md:30]: A full example with .things and .items file should be part of your readme.

File: bundles/org.openhab.binding.publictransportswitzerland/src/main/resources/OH-INF/binding/binding.xml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<binding:binding id="publictransportswitzerland" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:binding="https://openhab.org/schemas/binding/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Public Transport Switzerland Binding</name>
+	<description>Connects to the "Swiss public transport API" to provide real-time public transport information.</description>
+	<author>Jeremy Stucki &amp; Stefanie Jäger</author>
```
Comment (fwolter) [inline bundles/org.openhab.binding.publictransportswitzerland/src/main/resources/OH-INF/binding/binding.xml:8]: The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

File: bundles/org.openhab.binding.publictransportswitzerland/src/main/feature/feature.xml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<features name="org.openhab.binding.publictransportswitzerland-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
+
+	<feature name="openhab-binding-publictransportswitzerland" description="PublicTransportSwitzerland Binding" version="${project.version}">
```
Comment (fwolter) [inline bundles/org.openhab.binding.publictransportswitzerland/src/main/feature/feature.xml:5]: The description is displayed as binding name when installing it in MainUI.
```suggestion
	<feature name="openhab-binding-publictransportswitzerland" description="Public Transport Switzerland Binding" version="${project.version}">
```

**Review by fwolter — CHANGES_REQUESTED** [review]
Great to see progress on this!

There are some compiler warnings left. You could take a look at `mvn clean install`.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

File: bundles/org.openhab.binding.publictransportswitzerland/pom.xml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.2.0-SNAPSHOT</version>
```
Comment (fwolter) [inline bundles/org.openhab.binding.publictransportswitzerland/pom.xml:10]:
```suggestion
    <version>3.3.0-SNAPSHOT</version>
```

**Review by fwolter — CHANGES_REQUESTED** [review]

**Review by fwolter — APPROVED** [review]
LGTM

Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

**Comment by jeremystucki** [comment]
Thank you @fwolter for the fast and precise reviews, I'm so happy that it's gonna be in the next release :)

**Comment by lolodomo** [comment]
Please create a subsequent PR to add the default translation properties file. 
You can produce it simply by running this command : `mvn i18n:generate-default-translations`

**Pull request merged**